aggregator/src/datastore.rs

```diff
@@ -2885,6 +3070,155 @@ ORDER BY id DESC
+                    WHERE aggregation_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                      AND report_max_timestamps.max_timestamp < $2
+                      AND NOT EXISTS (
+                          SELECT id FROM collect_jobs
+                          WHERE aggregation_jobs.task_id = collect_jobs.task_id
+                            AND aggregation_jobs.batch_interval <@ collect_jobs.batch_interval)
+                ),
+                deleted_report_aggregations AS (
+                    DELETE FROM report_aggregations
+                    WHERE report_aggregations.aggregation_job_id IN (
+                        SELECT id FROM aggregation_jobs_to_delete)
+                ),
+                deleted_batch_aggregations AS (
+                    DELETE FROM batch_aggregations
+                    WHERE batch_aggregations.batch_identifier IN (
+                        SELECT batch_identifier FROM aggregation_jobs
+                        WHERE aggregation_jobs.id IN (SELECT id FROM aggregation_jobs_to_delete))
+                )
+                DELETE FROM aggregation_jobs
+                WHERE id IN (SELECT id FROM aggregation_jobs_to_delete)",
+            )
+            .await?;
+        self.tx
+            .execute(
+                &stmt,
+                &[
+                    /* task_id */ &task_id.get_encoded(),
+                    /* client_timestamp */
+                    &oldest_allowed_report_timestamp.as_naive_date_time()?,
+                ],
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Deletes old collection artifacts (collect jobs/aggregate share jobs/outstanding batches) for
+    /// a given task, that is, collection artifacts for which all associated client reports have
+    /// timestamps older than a given timestamp.
+    ///
+    /// Only implemented for leader tasks with the time-interval query type.
+    ///
+    /// After calling this function, delete_old_aggregation_artifacts must be called in the same
+    /// transaction to avoid re-collecting old aggregations.
+    #[tracing::instrument(skip(self), err)]
+    pub async fn delete_old_collect_artifacts(
+        &self,
+        task_id: &TaskId,
+        oldest_allowed_report_timestamp: Time,
+    ) -> Result<(), Error> {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "WITH collect_jobs_to_delete AS (
+                    SELECT id FROM collect_jobs
+                    JOIN (
+                        SELECT
+                            collect_jobs.id AS collect_job_id,
+                            MAX(client_reports.client_timestamp) AS max_timestamp
+                        FROM collect_jobs
+                        JOIN aggregation_jobs
+                            ON aggregation_jobs.task_id = collect_jobs.task_id
+                           AND aggregation_jobs.batch_interval <@ collect_jobs.batch_interval
```

Comment:
Continuing from my suggestion in `delete_old_aggregation_artifacts`, it seems like if `aggregation_jobs` had a `max_report_timestamp` column, we could avoid two JOINs here.

Reply:
Same as in that comment -- this would imply a schema change/data backfill, and I'd prefer to do denormalizations once we have data showing that they will be helpful.